# **Exercise 1**
Global variables



In [ ]:
def f(alist):
  list2=alist[0:len(alist)]
  for i in range(5):
      list2.append(i)
  return list2

alist = [1, 2, 3]
ans = f(alist)
print(ans)
print(alist) 

[1, 2, 3, 0, 1, 2, 3, 4]
[1, 2, 3]


# **Exercise 2**
List comprehension

In [1]:
theComprehension = [x*x for x in filter(lambda x: x % 2 == 1, range(10))]

print(theComprehension)

[1, 9, 25, 49, 81]


# **Exercie 3**

In [ ]:
number = int(input())
listin = input().split(' ')
def filterlongword(elementoflist):
  return len(elementoflist) <= number
print(list(filter(filterlongword,listin)))

3
kjnkjn hg
['hg']


# **Exercise 4**

In [ ]:
lang = {"Python" : 3, "Java" : '', "Cplusplus" : 'test', "Php" : 0.7}

def Keysofdictionary(dictionary):
  return list(map(len,dictionary.keys()))

print(Keysofdictionary(lang))

[6, 4, 9, 3]


# **Exercise 5**

In [ ]:
language_scores = [('Python', 97), ('Cplusplus', 81), ('Php', 45), ('Java', 32)]
language_scores.sort(key=lambda x:x[0])
print(language_scores)


[('Cplusplus', 81), ('Java', 32), ('Php', 45), ('Python', 97)]


# **Exercise 6**

In [ ]:
def Squarenum(number):
  return number**2
def Cubenum(number):
  return number**3
def Six(number):
  return(Cubenum(Squarenum(number)))
num = int(input())
print(Six(num))

3
729


# **Exercise 7**

In [ ]:
def hello(func):
    def wrapper(*args):
      print("Hello world!")
      return func(*args) 
    return wrapper

@hello
def square(x):
    return x*x

num = int(input())
print(square(num))

4
Hello world!
16


# **Exercise 8**

In [ ]:
def fibonacci(n):
    if(n <= 1):
        return n
    else:
        return(fibonacci(n-1) + fibonacci(n-2))


for i in range(1,21):
    print(fibonacci(i))

1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765


# **Exercise 9**

In [ ]:
import timeit
def loopFibonacci(n):
  fibonacci=[1,1]
  for i in range(0,n-2):
    fibonacci.append(fibonacci[i] + fibonacci[i+1])
  return fibonacci

def recursiveFibonacci(n):
    if(n <= 1):
        return n
    else:
        return(recursiveFibonacci(n-1) + recursiveFibonacci(n-2))



%timeit loopFibonacci(20)

%timeit recursiveFibonacci(20)

2.44 µs ± 42.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
1.91 ms ± 14.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


The loopFibonacci is better than recursiveFibonacci. The loopFibonacci is 2.44 microsecond with error of 42.6 nanos per loop but the recursiveFibonacci is 1.91 milisecond ± 14.7 micros per loop

# **Exercise 10**

In [6]:
import math

class Polygon:
    sidesofpoly = []
    def getandset(self, inpt):
      self.sidesofpoly = list(inpt.split(' '))

    def getOrderedSides(self, increasing):
        return sorted(self.sidesofpoly,key=lambda x:x[0], reverse =  not increasing)
 
    def perimeter(self):
      peri = 0
      for i in self.sidesofpoly:
        peri += int(i)
      return peri

poly = Polygon()
poly.getandset('7 8 9')
perimeterofpoly = poly.perimeter()
orderofpoly = poly.getOrderedSides(True)
print("The perimeter of a poly with sides of 7,8,9 is : ", perimeterofpoly,
      "\n The sides in order are :",orderofpoly)

The perimeter of a poly with sides of 7,8,9 is :  24 
 The sides in order are : ['9', '8', '7']


# **Exercise 11**

In [17]:
class Rectangle(Polygon):
  isRectangle = False
  def getandset(self, inpt):
      self.sidesofpoly = list(inpt.split(' '))
      if len(self.sidesofpoly) !=2:
        print("The polygon is not a rectangle.")
      else:
        self.isRectangle = True
        print("The polygon is a rectangle.")
  def area(self):
    if self.isRectangle:

      print("The area of the rectangle is : ", (int(self.sidesofpoly[0])*int(self.sidesofpoly[1])))

rect = Rectangle()
inpt = input()
rect.getandset(inpt)
rect.area()


3 4
The polygon is a rectangle.
The area of the rectangle is :  12
